### PR TITLE
Speedy Compiler: compute variable indices more generally.

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by cocreature — APPROVED** [review]
Very nice thank you! I didn’t review all the index changes in detail but I trust our tests to cach any typos there.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 1]
```diff
@@ -135,45 +135,56 @@ private[lf] final class Compiler(
...
+  def lookUpExprVar(name: ExprVarName): SEVar =
+    lookUpVar(EVarRef(name))
+      .getOrElse(throw CompilationError(s"Unknown variable: $name. Known: ${vars.mkString(",")}"))
+
+  def lookUpTypeVar(name: TypeVarName): Option[SEVar] =
+    lookUpVar(TVarRef(name))
+
+  private[this] case class Env(
+      position: Int = 0,
+      varIndices: Map[VarRef, Position] = Map.empty,
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:185]: I thinks the original idea was to represent a stack, and to be able to pop change made to it. 
In the current design this is not used, so better a `Map`.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 2]
```diff
@@ -135,45 +135,56 @@ private[lf] final class Compiler(
...
+
+  @inline
+  private[this] def nextPosition(): Position = {
+    val p = env.position
+    env = env.copy(position = env.position + 1)
+    Position(p)
+  }
 
+  @inline
+  private[this] def nextPosition(name: ExprVarName): Position = {
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala]: I dropped the overload.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 3]
```diff
@@ -135,45 +135,56 @@ private[lf] final class Compiler(
...
+
+  private[this] def addTypeVar(name: TypeVarName, position: Position) =
+    addVar(TVarRef(name), position)
+
+  private[this] def hideTypeVar(name: TypeVarName) =
+    env = env.copy(varIndices = env.varIndices - TVarRef(name))
+
+  private[this] def vars: List[VarRef] = env.varIndices.keys.toList
+
+  private[this] def lookUpVar(varRef: VarRef): Option[SEVar] =
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala]: fixed.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 4]
```diff
@@ -690,13 +695,13 @@ private[lf] final class Compiler(
   private[this] def compileAbss(expr0: Expr, arity: Int = 0): SExpr =
     expr0 match {
       case EAbs((binder, typ @ _), body, ref @ _) =>
-        env = env.addExprVar(binder)
+        nextPosition(binder)
         compileAbss(body, arity + 1)
       case ETyAbs((binder, KNat), body) =>
-        env = env.addTypeVar(binder)
+        addTypeVar(binder, nextPosition())
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:701]: no more overload.

**Pull request merged**